r/adbcdrivermanager/R/utils.R

```diff
@@ -94,3 +94,102 @@ str.adbc_xptr <- function(object, ...) {
   str(env_proxy, ...)
   invisible(object)
 }
+
+
+#' Low-level pointer details
+#'
+#' - `adbc_xptr_move()` allocates a fresh R object and moves all values pointed
+#'   to by `x` into it. The original R object is invalidated by zeroing its
+#'   content. This is useful when returning from a function where
+#'   [lifecycle helpers][with_adbc] were used to manage the original
+#'   object.
+#' - `adbc_xptr_is_valid()` provides a means by which to test for an invalidated
+#'   pointer.
+#'
+#' @param x An 'adbc_database', 'adbc_connection', 'adbc_statement', or
+#'   'nanoarrow_array_stream'
+#'
+#' @return
+#' - `adbc_xptr_move()`: A freshly-allocated R object identical to `x`
+#' - `adbc_xptr_is_valid()`: Returns FALSE if the ADBC object pointed to by `x`
+#'   has been invalidated.
+#' @export
+#'
+#' @examples
+#' db <- adbc_database_init(adbc_driver_void())
+#' adbc_xptr_is_valid(db)
+#' db_new <- adbc_xptr_move(db)
+#' adbc_xptr_is_valid(db)
+#' adbc_xptr_is_valid(db_new)
+#'
+adbc_xptr_move <- function(x) {
+  if (inherits(x, "adbc_database")) {
+    .Call(RAdbcMoveDatabase, x)
+  } else if (inherits(x, "adbc_connection")) {
+    .Call(RAdbcMoveConnection, x)
+  } else if (inherits(x, "adbc_statement")) {
+    .Call(RAdbcMoveStatement, x)
+  } else if (inherits(x, "nanoarrow_array_stream")) {
```

Comment:
Is this class also the responsibility of this function? Do we even need the genericity of `adbc_xptr_move()`, perhaps exposing specialized `adbc_xptr_move_database()` and friends? Same for `_is_valid()` .